### PR TITLE
Handle stemming in NLU injection

### PR DESCRIPTION
**Pull request by ClemDoum**
# Handle stemming in NLU injection

## Problem
Currently if the custom entity parser uses stems, injected values are add with their stemmed version in the `CustomEntityParser`

## Solution
- Stems are added to the custom entity parser depending on the custom entity parser `CustomEntityParserUsage`
- Replaced the `inject_entity_values` function with a `NluInjector` class that adds more flexibility for testing (it lets us mocked the share resources used in the injection for instance)
- Added a `load_engine_shared_resources` function that loads a the shared resources from the NLU engine directory path
- Added a `from_resources_with_path` method to the `NLUEngine` struct to be able to load a NLU engine with mocked shared resources

## Fixed
- the bug with the NLU where the path to the builtin gazetteer parser was wrongly built

*Note*: this is a breaking change

### CODE REVIEW

File: snips-nlu-lib/src/injection/injection.rs [hunk 1]
```diff
@@ -204,32 +297,111 @@ fn get_nlu_engine_metadata<P: AsRef<Path>>(engine_dir: P) -> Result<NLUEngineInf
...
 }
 
+fn normalize_entity_value(
+    entity_values: Vec<InjectedValue>
+) -> Vec<GazetteerEntityValue> {
+    entity_values
+        .into_iter()
+        .map(|value| GazetteerEntityValue {
+            raw_value: normalize(&*value),
+            resolved_value: value.to_string(),
```
Comment (adrienball) [inline snips-nlu-lib/src/injection/injection.rs]: This `to_string()` can be removed

File: snips-nlu-lib/src/intent_classifier/log_reg_intent_classifier.rs [hunk 2]
```diff
@@ -6,14 +6,16 @@ use std::sync::Arc;
 use itertools::Itertools;
 use ndarray::prelude::*;
 use serde_json;
+use snips_nlu_ontology::IntentClassifierResult;
 
 use models::IntentClassifierModel;
 use errors::*;
+use nlu_engine::load_engine_shared_resources;
```
Comment (adrienball) [inline snips-nlu-lib/src/intent_classifier/log_reg_intent_classifier.rs]: This import is only used in test, so we should move it there.

File: snips-nlu-lib/src/injection/injection.rs [hunk 3]
```diff
@@ -19,91 +29,175 @@ fn normalize(s: &str) -> String {
...
+        }?;
+
+        let maybe_stemmer = shared_resources.stemmer.as_ref();
+
+        // Normalize and stem all values if needed
+        info!("Normalizing injected values...");
+        let normalized_entity_values = self.entity_values
+            .into_iter()
+            .map(|(entity, values)| {
+                let normalized_values = normalize_entity_value(values);
```
Comment (adrienball) [inline snips-nlu-lib/src/injection/injection.rs]: `normalize_entity_values` would more consistent no ?

File: snips-nlu-lib/src/injection/injection.rs [hunk 4]
```diff
@@ -19,91 +29,175 @@ fn normalize(s: &str) -> String {
...
+                if engine_info.custom_entities.contains(&*entity) {
+                    let stemmed_normalized_values = stem_entity_value(
```
Comment (adrienball) [inline snips-nlu-lib/src/injection/injection.rs]: Same here, I'd rather use `stem_entity_values`

File: snips-nlu-lib/src/injection/injection.rs [hunk 5]
```diff
@@ -204,32 +297,111 @@ fn get_nlu_engine_metadata<P: AsRef<Path>>(engine_dir: P) -> Result<NLUEngineInf
...
+}
+
+fn stem_entity_value(
+    entity_values: Vec<GazetteerEntityValue>,
+    engine_info: &NluEngineInfo,
+    custom_entity_parser_info: &CustomGazetteerParserInfo,
+    maybe_stemmer: &Option<&Arc<Stemmer>>,
+) -> Result<Vec<GazetteerEntityValue>, NluInjectionError> {
+    let stemmed_values = match custom_entity_parser_info.parser_usage {
+        CustomEntityParserUsage::WithoutStems => Ok(vec![]),
```
Comment (adrienball) [inline snips-nlu-lib/src/injection/injection.rs]: I think you can just return `vec![]` here, and remove the `Ok(...)` in the other arm as well.

File: snips-nlu-lib/src/injection/injection.rs [hunk 6]
```diff
@@ -204,32 +297,111 @@ fn get_nlu_engine_metadata<P: AsRef<Path>>(engine_dir: P) -> Result<NLUEngineInf
...
+                        .map(|token| stemmer.stem(&*token))
+                        .join(" ");
+
+                    Ok(GazetteerEntityValue {
+                        raw_value,
+                        resolved_value: value.resolved_value.clone(),
+                    })
+                })
+                .collect::<Result<_, NluInjectionError>>()?;
+            Ok(stemmed)
```
Comment (adrienball) [inline snips-nlu-lib/src/injection/injection.rs]: You can remove `Ok(stemmed)` and directly return the expression before

File: snips-nlu-lib/src/intent_classifier/log_reg_intent_classifier.rs (hunk 2 above)
Comment (ClemDoum) [reply snips-nlu-lib/src/intent_classifier/log_reg_intent_classifier.rs]: it's used in the `inject` method

File: snips-nlu-lib/src/injection/injection.rs (hunk 5 above)
Comment (ClemDoum) [reply snips-nlu-lib/src/injection/injection.rs]: The other arms returns a result because the `stemmer` can be missing, I need to return a `Result`

**Pull request merged**